sklearn/cluster/kernel_kmeans.py

```diff
@@ -0,0 +1,279 @@
+        Distance update.
+
+    X_fit_ : array-like, shape=(n_samples, n_features)
+        Data used in clustering.
+
+    n_iter_ : Iteration in which algorithm converged
+    """
+
+    def __init__(self, n_clusters=3, max_iter=300, tol=1e-3, n_init=10,
+                 kernel='linear', gamma='auto', degree=3, coef0=1.0,
+                 kernel_params=None, random_state=None, verbose=0):
+        self.n_clusters = n_clusters
+        self.max_iter = max_iter
+        self.tol = tol
+        self.n_init = n_init
+        self.random_state = random_state
+        self.kernel = kernel
+        self.gamma = gamma
+        self.degree = degree
+        self.coef0 = coef0
+        self.kernel_params = kernel_params
+        self.verbose = verbose
+
+    def _check_fit_data(self, X):
+        """Verify that the number of samples given is larger than k"""
+        X = check_array(X, accept_sparse='csr', dtype=np.float64)
+        if X.shape[0] < self.n_clusters:
+            raise ValueError("n_samples=%d should be >= n_clusters=%d" % (
+                X.shape[0], self.n_clusters))
+        return X
+
+    def _check_test_data(self, X):
+        X = check_array(X, accept_sparse='csr', dtype=FLOAT_DTYPES,
+                        warn_on_dtype=True)
+        n_samples, n_features = X.shape
+        expected_n_features = self.X_fit_.shape[1]
+        if not n_features == expected_n_features:
+            raise ValueError("Incorrect number of features. "
+                             "Got %d features, expected %d" % (
+                                 n_features, expected_n_features))
+        return X
+
+    @property
+    def _pairwise(self):
+        return self.kernel == "precomputed"
+
+    def _get_kernel(self, X, Y=None):
+        if callable(self.kernel):
+            params = self.kernel_params or {}
+        else:
+            params = {"gamma": self.gamma,
+                      "degree": self.degree,
+                      "coef0": self.coef0}
+        return pairwise_kernels(X, Y, metric=self.kernel,
+                                filter_params=True, **params)
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Compute k-means clustering.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+
+        sample_weight : array-like, shape=(n_samples,)
+        """
+        X = self._check_fit_data(X)
+        n_samples = X.shape[0]
+
+        if self.n_init <= 0:
+            raise ValueError("Invalid number of initializations."
+                             " n_init=%d must be bigger than zero."
+                             % self.n_init)
+
+        if self.max_iter <= 0:
+            raise ValueError("Invalid iteration bound."
+                             " max_iter=%d must be greater than zero."
+                             % self.max_iter)
+
+        best_labels, best_inertia, best_distances = None, None, None
+
+        K = self._get_kernel(X)
```

Comment:
Please check that if no `gamma` is specified that the 'auto' string isn't being sent to the kernel operator. This is currently causing an issue.